### PR TITLE
[#9255] Migrate Instructor Course Enroll page to Angular (Part 1)

### PR DESCRIPTION
**Pull request by tanhengyeow**
Part of #9255 

Note:
I've combined `InstructorCourseEnrollSaveAction` and `InstructorCourseEnrollResult` because I intend to show them both in the same page (`instructor-course-enroll-page.component.html`). The content of `InstructorCourseEnrollResult` will be shown after a successful enroll.

Summary of PR:
- 2 API endpoints created:
  - `/courses/enrollSave` - Performs enrollment of students. Requires `courseid` and request body containing a formatted`string` of enrolled students. Returns `enrollResultPanelList`.
  - `/course/enroll/pageData` - Checks if the course id is valid before accessing the enroll page. Requires `courseid`. Returns `bool` and warning status messages (if any)
- Front-end:
  - Main enroll page migrated with `New students spreadsheet`.
  - Instructor course enroll results page is shown below the main enroll page, replacing the `More Info` part after a successful enroll. Failed enrolment is shown through status message like before.
  - Instructor course enroll results page will be phased out and be replaced with success/error messages shown on the respective spreadsheet rows instead.
  - **Note:** 
    - What was done initially was to not show the `enroll page with the spreadsheet` when the enroll results were shown. However, the HOT instance would be destroyed when the div of the `enroll page with the spreadsheet` is not shown in the page anymore. Since the Instructor course enroll results page will be phased out (refer to point above), stop gap solution is to show the results page below the spreadsheet interface after a successful enroll.
    - Clicking `Scroll down` when the enroll results page is shown will not show the `More Info` section. This will be ignored (refer to point above).

Todo:
- [x] Migrate back-end files/back-end tests
- [x] Migrate main enroll page with `New Students` spreadsheet
- [x] Migrate Instructor Course Enroll Results 
- [ ] Migrate `Existing students spreadsheet` (possibly in the next release?)
- [ ] Delete deprecated files after review

### CODE REVIEW

File: src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        String studentsInfo = getRequestBody();
+
+        /* Process enrollment list and setup data for page result */
+        try {
+            List<StudentAttributes>[] students = enrollAndProcessResultForDisplay(studentsInfo, courseId);
+            EnrollResults dataFormat = new EnrollResults(getInstructorCourseEnrollResult(students));
+            return new JsonResult(dataFormat);
+
+        } catch (EntityDoesNotExistException e) {
+            throw new EntityNotFoundException(new EntityDoesNotExistException(e));
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java]: Why not just `throw new EntityNotFoundException(e)`?

File: src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    private List<EnrollResultPanel> getInstructorCourseEnrollResult(List<StudentAttributes>[] students) {
+        List<EnrollResultPanel> enrollResultPanelList = new ArrayList<>();
+
+        for (int i = 0; i < StudentUpdateStatus.STATUS_COUNT; i++) {
+            String panelClass = "";
+
+            switch (StudentUpdateStatus.enumRepresentation(i)) {
+            case ERROR :
+                panelClass = "bg-danger";
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java]: For better separation of concern, let's not have HTML classes in Java code anymore.

File: src/web/app/components/status-message/status-message.component.html [hunk 3]
```diff
@@ -2,5 +2,7 @@
   <button class="close" (click)="dismiss(statusMessage)">
     <span aria-hidden="true">&times;</span>
   </button>
-  <div style="padding-right: 30px;">{{ statusMessage.message }}</div>
+
+  <!--TODO: Change this behavior when no status messages are returned in raw HTML-->
```
Comment (wkurniawan07) [inline src/web/app/components/status-message/status-message.component.html]: No need to have this TODO. Supporting status message with raw HTML may be something we actually need to support.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 4]
```diff
@@ -12,13 +67,205 @@ import { ActivatedRoute } from '@angular/router';
...
+
+    const hotInstanceColHeaders: string[] = (newStudentsHOTInstance.getColHeader() as string[]);
+    const dataPushToTextarea: string =
+        this.getUpdatedHeaderString(hotInstanceColHeaders);
+    const userDataRows: string = this.getUserDataRows(spreadsheetData);
+
+    this.enrollData = (userDataRows === ''
+        ? '' : dataPushToTextarea + userDataRows); // only include header string if userDataRows is not empty
+
+    const paramMap: { [key: string]: string } = { courseid: this.courseid };
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: `user: this.user` to support masquerade mode.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 5]
```diff
@@ -12,13 +67,205 @@ import { ActivatedRoute } from '@angular/router';
...
+        'insert_row', [], numOfRows);
+  }
+
+  /**
+   * Toggles the view of spreadsheet interface
+   * and/or its affiliated buttons
+   */
+  togglePanel(event: any): void {
+    this.targetElement =
+        event.target.closest('div').querySelector('i');
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Is there a way that's not dependent on DOM structure? Direct DOM manipulation is strongly discouraged in Angular.

File: src/web/app/pages-instructor/instructor-course-student-edit-page/instructor-course-student-edit-page.component.ts [hunk 6]
```diff
@@ -141,7 +141,7 @@ export class InstructorCourseStudentEditPageComponent implements OnInit, OnDestr
   /**
    * Sets the boolean value of `isSessionSummarySendEmail` to true if
    * user chooses to resend past session link to the new email.
-   * Submits the form otherwise.
+   * Always submits the form.
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-student-edit-page/instructor-course-student-edit-page.component.ts]: i.e. this comment can be removed entirely?

File: src/web/styles.scss [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 @import '~bootstrap/dist/css/bootstrap.css';
 @import '~@fortawesome/fontawesome-free/css/all.css';
+@import '~handsontable/dist/handsontable.full.css';
```
Comment (wkurniawan07) [inline src/web/styles.scss]: Does this need to be under global style? It's only used in the enroll page.

File: src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java (hunk 2 above)
Comment (tanhengyeow) [reply src/main/java/teammates/ui/newcontroller/PostCourseEnrollSaveAction.java]: @wkurniawan07 Let's change them to strings e.g. `danger`, `warning` and then apply a pipe to transform it to a css class in [here](https://github.com/TEAMMATES/teammates/pull/9329/files/291cc70d201d0c5ee3d115b0acfb6fdde48310b7#diff-35b02ff0aa8eefda8e49074c5ed8fc94R237)?

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts (hunk 5 above)
Comment (tanhengyeow) [reply src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Let me figure something out using `ngClass`.

**Pull request merged**